core/unix/signal.c

```diff
@@ -6693,6 +6700,141 @@ execute_native_handler(dcontext_t *dcontext, int sig, sigframe_rt_t *our_frame)
+    // be on it (because DR sets SA_ONSTACK for all signals, and that config is
+    // still in effect), regardless of whether the signal stack was registered
+    // for sig. If the signal stack was actually not registered for sig, then we
+    // return true (indicating that DR should continue trying to deliver the
+    // signal).
+    if (dcontext != NULL || !doing_detach)
+        return false;
+
+    bool has_sigstack = false;
+#ifdef HAVE_SIGALTSTACK
+    stack_t app_sigstack;
+    IF_DEBUG(int rc =)
+    sigaltstack_syscall(NULL, &app_sigstack);
+    ASSERT(rc == 0);
+    if (APP_HAS_SIGSTACK_EX(app_sigstack)) {
+        has_sigstack = true;
+        // If the app has a sigstack, we must be on it.
+        ASSERT(app_sigstack.ss_sp < (void *)xsp &&
+               (void *)xsp < app_sigstack.ss_sp + app_sigstack.ss_size);
+        // In release build, we bail on this optimization and try to deliver the
+        // signal.
+        return false;
+    }
+#endif
+
+    kernel_sigaction_t sigact_struct;
+    d_r_read_lock(&detached_sigact_lock);
+    memcpy(&sigact_struct, &detached_sigact[sig], sizeof(sigact_struct));
+    d_r_read_unlock(&detached_sigact_lock);
+
+#ifdef HAVE_SIGALTSTACK
+    // Satisfy "will never be NULL" complain from the compiler.
+    kernel_sigaction_t *sigact = &sigact_struct;
+    if (has_sigstack && !USE_APP_SIGSTACK_EX(app_sigstack, sigact)) {
+        // We are on the sigstack but we don't want to use it for signal delivery.
+        return false;
+    }
+#endif
+
+    if (sigact_struct.handler == (handler_t)SIG_IGN ||
+        sigact_struct.handler == (handler_t)SIG_DFL) {
+        // We handle logs, messages and asserts related to these cases in
+        // execute_native_handler() later.
+        return false;
+    }
+
+#ifdef X86_64
+    RSTATS_INC(num_signals);
+    RSTATS_INC(num_native_signals);
+
+    sigframe_rt_t *frame = (sigframe_rt_t *)xsp;
+    kernel_ucontext_t *uc = get_ucontext_from_rt_frame(frame);
+    sigcontext_t *sc = SIGCXT_FROM_UCXT(uc);
+    report_app_problem(
+        /*dcontext=*/NULL, APPFAULT_FAULT, (byte *)sc->SC_XIP, (byte *)sc->SC_FP,
+        "\nSignal %d delivered to application handler reusing current frame.\n", sig);
+
+    kernel_sigset_t blocked = sigact_struct.mask;
+    if (!TEST(SA_NOMASK, sigact_struct.flags))
+        kernel_sigaddset(&blocked, sig);
+
+    // Emulate the app's sigmask for this signal.
+    sigprocmask_syscall(SIG_SETMASK, &blocked, NULL, sizeof(blocked));
+    void (*asm_jmp_tgt)() = SIGACT_PRIMARY_HANDLER(&sigact_struct);
+    kernel_siginfo_t *siginfo_var = &((sigframe_rt_t *)xsp)->info;
+    kernel_ucontext_t *ucontext_var = &((sigframe_rt_t *)xsp)->uc;
+    asm volatile("mov %[jmp_tgt],  %%rcx\n\t"
+                 "mov     %[sig],  %%rdi\n\t"
+                 "mov %[siginfo],  %%rsi\n\t"
+                 "mov %[ucontext], %%rdx\n\t"
+                 "mov %[target_sp], %%rsp\n\t"
+                 "jmp *%%rcx\n\t"
```

Comment:
Do we need to create this entire alternate method of invoking the app's handler, which is x86-only?  I thought we would use the existing code in execute_native_handler() and just eliminate the copy of a new frame if the conditions of the stack and RT are met.  Isn't that simpler, more cross-arch, and avoids duplicate handling?

Reply:
execute_native_handler() redirects control to the native signal handler by setting up xip to the handler's address and setting up other regs so that the sigreturn from DR's main_signal_handler would cause the app's handler to be executed.

I tried something similar but concluded that we cannot let a sigreturn redirect control for us (since I saw crashes), because we want the app signal handler to operate on same the signal frame and need it to stay around.

Reply:
If there were something fundamental, that should be explained in comments so it's clear why this is using its own separate seemingly-duplicated attempt to invoke the native handler.  But I would expect using the sigreturn to work, since I don't think the kernel does much there but restore the machine state: so if the restored stack pointer points to the right place it should re-use the frame.

Reply:
To do it the sigreturn way, we'd need to clobber xdi, xsi, and xdx in the sigcontext, and yet the same sigcontext will be used for the native signal delivery; this is a transparency issue and may cause problems for the app if the native signal handler wants to examine those values https://github.com/DynamoRIO/dynamorio/blob/6aa787349c98cca0befd5a831b0a2de542c7a9d7/core/unix/signal.c#L6653

Reply:
Ah, ok, so need to have the frame's state be the interrupted state yet also set the parameters to the app's handler using the same state.  We could target a trampoline that sets the registers?  I just see a lot of logic in the execute_native_handler() and want to avoid building up a whole parallel routine where it's easy to miss a check: if someone adds a check to execute_native_handler() they'd have to remember to go update this other parallel routine.  That's why it seems worth trying to route everything through the same code.  If a trampline is too hard to set up: could we still share the other code and just differ in the control xfer?